Python/bltinmodule.c

```diff
@@ -2737,6 +2739,496 @@ builtin_issubclass_impl(PyObject *module, PyObject *cls,
     return PyBool_FromLong(retval);
 }
 
+/*[clinic input]
+isimmutable as builtin_isimmutable
+
+    obj: object
+    /
+
+Return whether 'obj' is immutable.
+[clinic start generated code]*/
+
+static PyObject *
+builtin_isimmutable(PyObject *module, PyObject *obj)
+/*[clinic end generated code: output=80c746a3bd7adb46 input=15c0c9d2da47bc15]*/
+{
+    _Py_VPYDBG("isimmutable(");
+    _Py_VPYDBGPRINT(obj);
+    _Py_VPYDBG(") region: %lu\n", Py_REGION(obj));
+    return PyBool_FromLong(_Py_IsImmutable(obj));
+}
+
+typedef struct node_s {
+    PyObject* object;
+    struct node_s* next;
+} node;
+
+typedef struct stack_s {
+    node* head;
+} stack;
+
+stack* stack_new(void){
+    stack* s = (stack*)malloc(sizeof(stack));
+    if(s == NULL){
+        return NULL;
+    }
+
+    s->head = NULL;
+
+    return s;
+}
+
+bool stack_push(stack* s, PyObject* object){
+    node* n = (node*)malloc(sizeof(node));
+    if(n == NULL){
+        return true;
+    }
+
+    n->object = object;
+    Py_INCREF(object);
+    n->next = s->head;
+    s->head = n;
+    return false;
+}
+
+PyObject* stack_pop(stack* s){
+    if(s->head == NULL){
+        return NULL;
+    }
+
+    node* n = s->head;
+    PyObject* object = n->object;
+    s->head = n->next;
+    free(n);
+    Py_DECREF(object);
+
+    return object;
+}
+
+void stack_free(stack* s){
+    while(s->head != NULL){
+        stack_pop(s);
+    }
+
+    free(s);
+}
+
+bool stack_empty(stack* s){
+    return s->head == NULL;
+}
+
+void stack_print(stack* s){
+    _Py_VPYDBG("stack: ");
+    node* n = s->head;
+    while(n != NULL){
+        _Py_VPYDBGPRINT(n->object);
+        _Py_VPYDBG("[rc=%ld]\n", n->object->ob_refcnt);
+        n = n->next;
+    }
+}
+
+bool is_leaf(PyObject* obj){
+    return Py_IsNone(obj) || Py_IsTrue(obj) || Py_IsFalse(obj) || PyUnicode_Check(obj) || PyLong_Check(obj) || PyBytes_Check(obj) || PyFrozenSet_Check(obj);
+}
+
+bool is_c_wrapper(PyObject* obj){
+    return PyCFunction_Check(obj) || Py_IS_TYPE(obj, &_PyMethodWrapper_Type) || Py_IS_TYPE(obj, &PyWrapperDescr_Type);
+}
+
+bool is_leaf_type(PyTypeObject* obj){
+    return PyType_HasFeature(obj, Py_TPFLAGS_IMMUTABLETYPE) || !PyType_HasFeature(obj, Py_TPFLAGS_HEAPTYPE);
+}
+
+PyObject* add_elements_fast(PyObject* seq, stack* frontier) {
+    Py_ssize_t size;
+    _PyObject_ASSERT(seq, PyList_Check(seq) || PyTuple_Check(seq));
+    size = PySequence_Fast_GET_SIZE(seq);
+    _Py_VPYDBG("list/tuple: pushing %ld [i: element] elements\n", size);
+    for(Py_ssize_t i = 0; i < size; i++){
+        PyObject* element = PySequence_Fast_GET_ITEM(seq, i); // element.rc = x
+        _Py_VPYDBG("[%ld: ", i);
+        _Py_VPYDBGPRINT(element);
+        _Py_VPYDBG("]\n");
+        if(!_Py_IsImmutable(element)){
+            Py_INCREF(element); // element.rc = x + 1
+            if(stack_push(frontier, element)){
+                Py_DECREF(element);
+                return PyErr_NoMemory();
+            }
```

Comment:
`stack_push` performs an IncRef in the successful case.  Can we write this just as
```suggestion
            if(stack_push(frontier, element)){
                return PyErr_NoMemory();
            }
```